base/sparse/higherorderfns.jl

```diff
@@ -966,99 +976,61 @@ broadcast(f::Tf, A::SparseMatrixCSC, ::Type{T}) where {Tf,T} = broadcast(x -> f(
-promote_containertype(::Type{PromoteToSparse}, ::Type{Array}) = PromoteToSparse
-promote_containertype(::Type{Array}, ::Type{PromoteToSparse}) = PromoteToSparse
-# combinations involving Arrays and sparse arrays continue in the promote-to-sparse funnel
-promote_containertype(::Type{AbstractSparseArray}, ::Type{Array}) = PromoteToSparse
-promote_containertype(::Type{Array}, ::Type{AbstractSparseArray}) = PromoteToSparse
+PromoteToSparse(::Val{0}) = PromoteToSparse()
+PromoteToSparse(::Val{1}) = PromoteToSparse()
+PromoteToSparse(::Val{2}) = PromoteToSparse()
+PromoteToSparse(::Val{N}) where N = Broadcast.DefaultArrayStyle{N}()
 
-# second (internal sparse broadcast containertype) dispatch layer's promotion logic
-# mostly just disambiguates Array from the main containertype promotion mechanism
-# AbstractArray serves as a marker to shunt to the generic AbstractArray broadcast code
-_spcontainertype(x) = _containertype(x)
-_spcontainertype(::Type{<:Vector}) = Vector
-_spcontainertype(::Type{<:Matrix}) = Matrix
-_spcontainertype(::Type{<:RowVector}) = Matrix
-_spcontainertype(::Type{<:Ref}) = AbstractArray
-_spcontainertype(::Type{<:AbstractArray}) = AbstractArray
-# need the following two methods to override the immediately preceding method
-_spcontainertype(::Type{<:StructuredMatrix}) = PromoteToSparse
-_spcontainertype(::Type{<:SparseVecOrMat}) = AbstractSparseArray
-spcontainertype(x) = _spcontainertype(typeof(x))
-spcontainertype(ct1, ct2) = promote_spcontainertype(spcontainertype(ct1), spcontainertype(ct2))
-@inline spcontainertype(ct1, ct2, cts...) = promote_spcontainertype(spcontainertype(ct1), spcontainertype(ct2, cts...))
+Broadcast.BroadcastStyle(::PromoteToSparse, ::SPVM) = PromoteToSparse()
+Broadcast.BroadcastStyle(::PromoteToSparse, ::Broadcast.Style{Tuple}) = Broadcast.DefaultArrayStyle{2}()
 
-promote_spcontainertype(::Type{T}, ::Type{T}) where {T} = T
-# combinations involving AbstractArrays and/or Tuples divert to the generic AbstractArray broadcast code
-DivertToAbsArrayBC = Union{Type{AbstractArray},Type{Tuple}}
-promote_spcontainertype(::DivertToAbsArrayBC, ct) = AbstractArray
-promote_spcontainertype(ct, ::DivertToAbsArrayBC) = AbstractArray
-promote_spcontainertype(::DivertToAbsArrayBC, ::DivertToAbsArrayBC) = AbstractArray
-# combinations involving scalars, sparse arrays, structured matrices (PromoteToSparse),
-# dense vectors/matrices, and PromoteToSparse collections continue in the promote-to-sparse funnel
-FunnelToSparseBC = Union{Type{Any},Type{Vector},Type{Matrix},Type{PromoteToSparse},Type{AbstractSparseArray}}
-promote_spcontainertype(::FunnelToSparseBC, ::FunnelToSparseBC) = PromoteToSparse
+Broadcast.BroadcastStyle(::SPVM, ::Broadcast.DefaultArrayStyle{0}) = PromoteToSparse()
+Broadcast.BroadcastStyle(::SPVM, ::Broadcast.DefaultArrayStyle{1}) = PromoteToSparse()
+Broadcast.BroadcastStyle(::SPVM, ::Broadcast.DefaultArrayStyle{2}) = PromoteToSparse()
```

Comment:
Do these definitions cause sparse broadcast to capture a broader array of types?

Reply:
Yes, deliberately. What are the obstacles to handling generic `AbstractArray`s of the proper dimensionality?

Reply:
https://github.com/JuliaLang/julia/pull/23939#discussion_r143014118 :)

> What are the obstacles to handling generic AbstractArrays of the proper dimensionality?

Soundly tested and with good performance? In principle, nothing. In practice, a large amount of work unlikely to happen any time soon :smile:.

Reply:
I've been thinking that for consistency they should at least work irrespectively of their performance (people will eventually complain and we can gracefully tell them that we accept PRs to fix it). But I haven't really looked into the cases that might fail and how hard it would be to at least guarantee them to work. If it's is not too hard I'd we should for it, but I suspect that is part of the problem here.

Reply:
Let's get the API we want for 1.0, and then fix performance/correctness bugs as they come up. If we add a restriction I'd be much more in favor of doing it based on `Base.OneTo` indices. I don't think that limiting it to `Array` makes very much sense, especially since recent improvements mean that people work with fewer and fewer plain-ol `Array`s.

Reply:
To clarify, I am not arguing against expansion of the set of types that participate in sparse broadcast. To the contrary, I agree that set of types likely should expand :).

Rather, now does not seem the right time to push that expansion: Sparse broadcast would benefit from quite a bit of work on a number of fronts. I would be delighted to see that work through. But with things sparse moving out of base, that work can happen post-1.0, including changing the set of types that participate in sparse broadcast. Simultaneously, there is much other work that *must* happen pre-1.0, and development resources are limited. As such, allocating those scarce resources to sparse broadcast pre-1.0 seems unwise, and expanding the set of types that participate in sparse broadcast increases pressure to do so. Best!

Reply:
I was (and am) fine with this, but it turns out to have unfortunate implications even for 1.x https://github.com/JuliaArrays/StaticArrays.jl/pull/348#discussion_r155953668.